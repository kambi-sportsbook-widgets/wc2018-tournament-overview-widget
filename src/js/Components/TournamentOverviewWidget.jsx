import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { ScrolledList, BlendedBackground } from 'kambi-widget-components'

import mobile from '../Services/mobile'
import kambi from '../Services/kambi'
import styles from './TournamentOverviewWidget.scss'
import { widgetModule, translationModule } from 'kambi-widget-core-library'

import Event from './Event'
import ArrowButton from './ArrowButton'
import ItemContainer from './ItemContainer'
import TournamentLogo from './TournamentLogo'
import List from './List'
import ListItem from './ListItem'

/**
 * How long (in milliseconds) to wait before scrolling league logo out
 * @type {number}
 */
const MOBILE_INITIAL_SCROLL_DELAY = 2000
const WORLD_CUP_2018_ID = 2000075007
const WIDGET_HEIGHT_DESKTOP = 350
const WIDGET_HEIGHT_MOBILE = 150
const DEFAULT_BACKGROUND = 'assets/overview-bw-bg-desktop.jpg'

const t = translationModule.getTranslation.bind(translationModule)

class TournamentOverviewWidget extends Component {
  /**
   * Constructs.
   * @param {object} props Component properties
   */
  constructor(props) {
    super(props)

    this.state = {
      events: props.events,
      selected: 0,
      mobile: mobile(),
      widgetHeight: mobile() ? WIDGET_HEIGHT_MOBILE : WIDGET_HEIGHT_DESKTOP,
      usingDefaultBackground: props.backgroundUrl === DEFAULT_BACKGROUND
    }
  }

  /**
   * Called after component rendering to DOM.
   */
  componentDidMount() {
    if (mobile()) {
      setTimeout(
        () => this.setState({ selected: 1 }),
        MOBILE_INITIAL_SCROLL_DELAY
      )
    }
    widgetModule.setWidgetHeight(this.state.widgetHeight)
  }

  componentDidUpdate() {
    widgetModule.setWidgetHeight(this.state.widgetHeight)
  }


  /**
   * Generates country icon url
   * country { string } country name e.g 'Sweden', 'South Korea'
   */
  generateCountryFlagUrl(country) {
    const normalisedCountryName = country.toLowerCase().replace(/\s/g, '_')
    return `${this.props.flagBaseUrl}${normalisedCountryName}.svg`
  }

  /**
   * Navigates to event
   * event { shape } event containing betOffer and event metadata
   */
  navigateToEvent(event) {
    if (event.event.liveBetOffers) {
      // Navigate to live event
      // http://kambi-sportsbook-widgets.github.io/widget-core-library/module-widgetModule.html#.navigateToLiveEvent__anchor
      widgetModule.navigateToLiveEvent(event.event.id)
    } else {
      // Navigate to upcoming event
      // http://kambi-sportsbook-widgets.github.io/widget-core-library/module-widgetModule.html#.navigateToEvent__anchor
      widgetModule.navigateToEvent(event.event.id)
    }
  }

  /**
   * Generates title for each widget item from event name
   * event { shape } contains event metadata
   */
  generateWidgetItemTitle(event) {
    return `${event.group} - ${event.name.split('(')[0].trim()}` // e.g "WM 2018 - Skyttekung"
  }

  /**
   * Sorts outcomes by lowest odds and label if same odds
   * outcomes { array } outcomes to sort
   */
  sortOutcomesByLowestOdds(outcomes) {
    return outcomes.sort((a, b) => {
      if (a.odds < b.odds) {
        return -1;
      } else if (a.odds === b.odds) {
        if (a.label < b.label) {
          return -1;
        }
        return 1;
      } else if (a.odds > b.odds) {
        return 1;
      }
    });
  }

  /**
   * Returns event node to display at top of widget
   * eventData { shape } eventData object containing betOffer and event metadata
   * numberOfOutcomes { number } how many outcomes we want the widget to render
   */
  renderTopEvent(eventData, numberOfOutcomes = 3) {
    return (
      <List
        title={this.generateWidgetItemTitle(eventData.event)}
        handleClick={() => this.navigateToEvent(eventData)}
        showNavLink={eventData.betOffers[0].outcomes.length > numberOfOutcomes}
        navText={t('showAllParticipants', eventData.betOffers[0].outcomes.length)}
      >
        {
          this.sortOutcomesByLowestOdds(eventData.betOffers[0].outcomes, numberOfOutcomes)
            .slice(0, numberOfOutcomes)
            .map(outcome => {
              let flagUrl = null
              const participant = outcome.label.split('(')[0]
              const countrySplit = outcome.englishLabel.split('(')
              
              if (countrySplit && countrySplit.length > 1 && eventData.event.groupId === WORLD_CUP_2018_ID) {
                flagUrl = this.generateCountryFlagUrl(countrySplit[1].slice(0, countrySplit[1].length -1))
              } else if (eventData.event.groupId === WORLD_CUP_2018_ID) {
                flagUrl = this.generateCountryFlagUrl(countrySplit[0])
              }

              return (
                <ListItem
                  key={outcome.id}
                  participant={participant}
                  flagUrl={flagUrl}
                  fallbackFlagUrl={'../../assets/world_cup_2018.svg'}
                  outcome={outcome}
                  event={eventData.event}
                />
          )})
        }            
      </List>
    )
  }

  /**
   * Renders component.
   * @returns {XML}
   */
  render() {
    const { leftWidget, rightWidget } = this.props.competitions
    
    return (
      <div className={[styles.widget, 'KambiWidget-primary-background-color'].join(' ')}>
        <BlendedBackground backgroundUrl={this.props.backgroundUrl} blendWithOperatorColor={this.state.usingDefaultBackground} />
        {
          !mobile() &&
          <div className={styles.topArea}>
            { this.renderTopEvent(leftWidget) }
            <TournamentLogo logoUrl={this.props.iconUrl} />
            { this.renderTopEvent(rightWidget) }
          </div>
        }
        <ScrolledList
          renderPrevButton={props => <ArrowButton type="left" {...props} />}
          renderNextButton={props => <ArrowButton type="right" {...props} />}
          renderItemContainer={props => <ItemContainer {...props} />}
          selected={this.state.selected}
          scrollToItemMode={ScrolledList.SCROLL_TO_ITEM_MODE.TO_LEFT}
          showControls={!mobile()}
        >
          {/* { mobile() && <TournamentLogo logoUrl={this.props.iconUrl} />} */}
          {this.state.events
            .filter(event => event.betOffers.length > 0)
            .map(event => {
              const countries = event.event.englishName.split(' - ')
              const isWorldCup = event.event.groupId === WORLD_CUP_2018_ID
              return (
                <Event
                  key={event.event.id}
                  event={event.event}
                  liveData={event.liveData}
                  outcomes={event.betOffers[0].outcomes}
                  homeFlag={isWorldCup ? this.generateCountryFlagUrl(countries[0]) : null}
                  awayFlag={isWorldCup ? this.generateCountryFlagUrl(countries[1]) : null}
                />
              )
            })}
        </ScrolledList>
      </div>
    )
  }
}

/**
 * @property events {Array} events to display in the match overview bottom widget
 * @property compentitions {Array} competitions with keys leftWidget, rightWidget containing their respective betOffers
 * @property iconUrl {String} provides full path to tournament icon svg
 * @property backgroundUrl {String} provides path to backgroundImage
 */
TournamentOverviewWidget.propTypes = {
  events: PropTypes.arrayOf(PropTypes.object).isRequired,
  competitions: PropTypes.shape().isRequired,
  backgroundUrl: PropTypes.string,
  flagBaseUrl: PropTypes.string,
  iconUrl: PropTypes.string,
  eventsRefreshInterval: PropTypes.number,
  pollingCount: PropTypes.number,
  filter: PropTypes.string.isRequired
}

export default TournamentOverviewWidget
