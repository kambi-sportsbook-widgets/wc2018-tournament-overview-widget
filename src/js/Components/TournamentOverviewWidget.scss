.widget {
   color: white;
   position: relative;
   padding: 12px 0;
}

.topArea {
   display: flex;
   max-width: 850px;
   min-width: 650px;
   margin: 0 auto 1rem auto;
   justify-content: space-between;
   padding: 0 2rem;
}
