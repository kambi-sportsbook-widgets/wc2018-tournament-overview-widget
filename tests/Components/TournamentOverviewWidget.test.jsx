/* eslint-env jest */
import React from 'react'
import ReactShallowRenderer from 'react-test-renderer/shallow'
import { shallow } from 'enzyme'
import {
  widgetModule,
  coreLibrary,
  translationModule,
} from 'kambi-widget-core-library'

import TournamentOverviewWidget from '../../src/js/Components/TournamentOverviewWidget'

let renderer

const mockEvent = {
  id: 100,
  group: 'VM 2018',
  groupId: 393929,
  name: 'Mock name',
  englishName: 'Mock english name',
  liveData: {},
  betOffers: [{ id: 1, outcomes: [] }],
}

const mockCompetitions = {
  leftWidget: mockEvent,
  rightWidget: mockEvent,
}

const mockBgUrl = 'some/fake/path.jpg'

let mockMobile = false

jest.mock('../../src/js/Services/mobile', () => () => mockMobile)
jest.mock('kambi-widget-core-library', () => ({
  widgetModule: {
    setWidgetHeight: jest.fn(),
  },
  translationModule: {
    getTranslation: jest.fn(key => `Translated ${key}`),
  },
  coreLibrary: {
    rootElement: <div />,
  },
}))

jest.useFakeTimers()

const widget = (
  <TournamentOverviewWidget
    events={[mockEvent]}
    competitions={mockCompetitions}
    backgroundUrl={mockBgUrl}
  />
)

describe('TournamentOverviewWidget DOM rendering', () => {
  beforeEach(() => {
    renderer = new ReactShallowRenderer()
    mockMobile = false
    translationModule.getTranslation = jest.fn()
    widgetModule.setWidgetHeight = jest.fn()
    coreLibrary.rootElement = <div />
  })

  it('renders correctly with default props', () => {
    expect(renderer.render(widget)).toMatchSnapshot()
  })

  it('renders correctly with tournament logo', () => {
    expect(
      renderer.render(
        <TournamentOverviewWidget
          events={[mockEvent]}
          competitions={mockCompetitions}
          tournamentLogo="test_tournament_logo"
          backgroundUrl={mockBgUrl}
        />
      )
    ).toMatchSnapshot()
  })

  it('renders correctly with events', () => {
    expect(renderer.render(widget)).toMatchSnapshot()
  })

  it('renders correctly with events and betoffers', () => {
    expect(renderer.render(widget)).toMatchSnapshot()
  })

  it('renders correctly in desktop mode', () => {
    mockMobile = false
    const wrapper = shallow(widget)
    expect(wrapper.debug()).toMatchSnapshot()
  })

  it('renders correctly in mobile mode', () => {
    mockMobile = true

    const wrapper = shallow(widget)

    expect(wrapper.debug()).toMatchSnapshot()

    jest.runAllTimers()

    expect(wrapper.debug()).toMatchSnapshot()
  })
})
