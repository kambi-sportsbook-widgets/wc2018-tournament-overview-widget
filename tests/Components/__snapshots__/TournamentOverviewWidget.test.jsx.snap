// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TournamentOverviewWidget DOM rendering renders correctly in desktop mode 1`] = `
"<div className=\\"widget\\">
  <BlendedBackground backgroundUrl=\\"some/fake/path.jpg\\" blendWithOperatorColor={false} />
  <div className=\\"topArea\\">
    <List title=\\"VM 2018 - Mock name\\" handleClick={[Function: handleClick]} showNavLink={false} navText=\\"Translated showAllParticipants\\" />
    <TournamentLogo logoUrl={[undefined]} logoName=\\"football\\" />
    <List title=\\"VM 2018 - Mock name\\" handleClick={[Function: handleClick]} showNavLink={false} navText=\\"Translated showAllParticipants\\" />
  </div>
  <ScrolledList renderPrevButton={[Function: renderPrevButton]} renderNextButton={[Function: renderNextButton]} renderItemContainer={[Function: renderItemContainer]} selected={0} scrollToItemMode=\\"to-left\\" showControls={true} step={2} alignItems=\\"center\\">
    <Event event={{...}} liveData={{...}} outcomes={{...}} homeFlag={{...}} awayFlag={{...}} isMobile={false} flagBaseUrl=\\"assets/icons/\\" />
  </ScrolledList>
</div>"
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly in mobile mode 1`] = `
"<div className=\\"widget\\">
  <BlendedBackground backgroundUrl=\\"some/fake/path.jpg\\" blendWithOperatorColor={false} />
  <ScrolledList renderPrevButton={[Function: renderPrevButton]} renderNextButton={[Function: renderNextButton]} renderItemContainer={[Function: renderItemContainer]} selected={0} scrollToItemMode=\\"to-left\\" showControls={false} step={2} alignItems=\\"center\\">
    <TournamentLogo logoUrl={[undefined]} logoName=\\"football\\" />
    <Event event={{...}} liveData={{...}} outcomes={{...}} homeFlag={{...}} awayFlag={{...}} isMobile={true} flagBaseUrl=\\"assets/icons/\\" />
  </ScrolledList>
</div>"
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly in mobile mode 2`] = `
"<div className=\\"widget\\">
  <BlendedBackground backgroundUrl=\\"some/fake/path.jpg\\" blendWithOperatorColor={false} />
  <ScrolledList renderPrevButton={[Function: renderPrevButton]} renderNextButton={[Function: renderNextButton]} renderItemContainer={[Function: renderItemContainer]} selected={0} scrollToItemMode=\\"to-left\\" showControls={false} step={2} alignItems=\\"center\\">
    <TournamentLogo logoUrl={[undefined]} logoName=\\"football\\" />
    <Event event={{...}} liveData={{...}} outcomes={{...}} homeFlag={{...}} awayFlag={{...}} isMobile={true} flagBaseUrl=\\"assets/icons/\\" />
  </ScrolledList>
</div>"
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly with default props 1`] = `
<div
  className="widget"
>
  <BlendedBackground
    backgroundUrl="some/fake/path.jpg"
    blendWithOperatorColor={false}
  />
  <div
    className="topArea"
  >
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
    <TournamentLogo
      logoName="football"
      logoUrl={undefined}
    />
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
  </div>
  <ScrolledList
    alignItems="center"
    renderItemContainer={[Function]}
    renderNextButton={[Function]}
    renderPrevButton={[Function]}
    scrollToItemMode="to-left"
    selected={0}
    showControls={true}
    step={2}
  >
    <Event
      awayFlag={null}
      event={
        Object {
          "englishName": "Mock english name",
          "group": "VM 2018",
          "groupId": 393929,
          "id": 100,
          "name": "Mock name",
        }
      }
      flagBaseUrl="assets/icons/"
      homeFlag={null}
      isMobile={false}
      liveData={Object {}}
      outcomes={Array []}
    />
  </ScrolledList>
</div>
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly with events 1`] = `
<div
  className="widget"
>
  <BlendedBackground
    backgroundUrl="some/fake/path.jpg"
    blendWithOperatorColor={false}
  />
  <div
    className="topArea"
  >
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
    <TournamentLogo
      logoName="football"
      logoUrl={undefined}
    />
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
  </div>
  <ScrolledList
    alignItems="center"
    renderItemContainer={[Function]}
    renderNextButton={[Function]}
    renderPrevButton={[Function]}
    scrollToItemMode="to-left"
    selected={0}
    showControls={true}
    step={2}
  >
    <Event
      awayFlag={null}
      event={
        Object {
          "englishName": "Mock english name",
          "group": "VM 2018",
          "groupId": 393929,
          "id": 100,
          "name": "Mock name",
        }
      }
      flagBaseUrl="assets/icons/"
      homeFlag={null}
      isMobile={false}
      liveData={Object {}}
      outcomes={Array []}
    />
  </ScrolledList>
</div>
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly with events and betoffers 1`] = `
<div
  className="widget"
>
  <BlendedBackground
    backgroundUrl="some/fake/path.jpg"
    blendWithOperatorColor={false}
  />
  <div
    className="topArea"
  >
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
    <TournamentLogo
      logoName="football"
      logoUrl={undefined}
    />
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
  </div>
  <ScrolledList
    alignItems="center"
    renderItemContainer={[Function]}
    renderNextButton={[Function]}
    renderPrevButton={[Function]}
    scrollToItemMode="to-left"
    selected={0}
    showControls={true}
    step={2}
  >
    <Event
      awayFlag={null}
      event={
        Object {
          "englishName": "Mock english name",
          "group": "VM 2018",
          "groupId": 393929,
          "id": 100,
          "name": "Mock name",
        }
      }
      flagBaseUrl="assets/icons/"
      homeFlag={null}
      isMobile={false}
      liveData={Object {}}
      outcomes={Array []}
    />
  </ScrolledList>
</div>
`;

exports[`TournamentOverviewWidget DOM rendering renders correctly with tournament logo 1`] = `
<div
  className="widget"
>
  <BlendedBackground
    backgroundUrl="some/fake/path.jpg"
    blendWithOperatorColor={false}
  />
  <div
    className="topArea"
  >
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
    <TournamentLogo
      logoName="football"
      logoUrl={undefined}
    />
    <List
      handleClick={[Function]}
      navText="Translated showAllParticipants"
      showNavLink={false}
      title="VM 2018 - Mock name"
    />
  </div>
  <ScrolledList
    alignItems="center"
    renderItemContainer={[Function]}
    renderNextButton={[Function]}
    renderPrevButton={[Function]}
    scrollToItemMode="to-left"
    selected={0}
    showControls={true}
    step={2}
  >
    <Event
      awayFlag={null}
      event={
        Object {
          "englishName": "Mock english name",
          "group": "VM 2018",
          "groupId": 393929,
          "id": 100,
          "name": "Mock name",
        }
      }
      flagBaseUrl="assets/icons/"
      homeFlag={null}
      isMobile={false}
      liveData={Object {}}
      outcomes={Array []}
    />
  </ScrolledList>
</div>
`;
